content/post/fewer-parameters-means-fewer-bugs.md

```diff
@@ -0,0 +1,244 @@
+When I review code, I sometimes notice that there is an opportunity to squash method parameters - yet many developers are unaware of this when they write code. From my coding experience, we can write code more efficiently and reduce the amount of bugs by making an effort to reduce the number of parameters in our methods. Unless your manager still evaluates output by lines of code, you should definitely try these tips in your daily coding life.
+
+So today, I will share how to do it with "OOP" (object-oriented programming) design, and explain where this magic power comes from.
+
+Let's start with a small example. The following demo code was written in Ruby, but don't worry; Ruby code is so beginner-friendly that once you have some "OOP" background, it's easy to grasp.
+
+The example story is, we have some books and want to pass the author name as a parameter, and get the book title, author name, and year published as a summary.
+```ruby
+class Book
+  attr_reader :title, :year_published
+ 
+  def initialize(title, year_published)
+    @title = title
+    @year_published = year_published
+  end
+ 
+  def summary(first_name, last_name)
+    "#{title} - #{last_name}, #{first_name} (#{year_published})"
+  end
+end
+ 
+book = Book.new("Ruby is the best", "2020")
+book.summary("Bater", "Chen")
+=> "Ruby is the best - Chen, Bater (2020)"
+```
+We have a book class. It has two attributes: `title`, and `year_published`. The `initialize` method is called when we `new` a book object. Book class only has one instance method called `summary`. It needs two parameters: first_name and last_name. In Ruby, when we define the `attr_reader` as the attribute of the object, it will create the read method automatically. So we can call the instance variable directly by `title` to get `@title` return, same as `year_published`.
+
+This simple method works fine but is not good enough. Let's refactor it.
+
+## The good code should be small and simple
+What is good code? This question might be hard to answer, but I would say **good code should be small and simple.** In this article, let's focus on making the parameters smaller.
+
+Generally speaking, _**one** parameter is better than **two**_ and _**no** parameter is better than **one**._ The problem with parameters is that it takes time to read and understand them, and time is always expensive. We spent around 70% of time reading the existing code, but only 30% is spent actually writing. If we can save some time from reading the code, not only from literally reading but also from understanding the spec and logic, we can write code more efficiently.
+
+Secondly, **when code is short, bugs have nowhere to hide.** Parameters are bug-prone because it demands extra care and concern. Reducing parameters can not only make the method more clear, it also shows our intention explicitly. It can help us to maintain the code easily. When we reduce our parameters, we are actually reducing the number of potential bugs at the same time.
```

Comment:
A bit of self promo: this could do with a footnote pointing to https://nemo.oudeis.fyi/posts/make-it-explicit/ 😁